### ISSUE TITLE
Goals are not marked @threadSafe (Parallel Maven Build)

### ISSUE THREAD

**Issue by ciis0** [eclipse/xtext-maven#36]
Hello,

due to the long runtime our project has, we are currently trying to make it faster by building it in parallel.

But, as the warning below tells, this plugin is not marked as thread-safe so I am reporting this issue to kindly request thread-safety. :)

Or, if this is already known internally, maybe just point me where thread-safety has to established and I could look into that.

Thank you.

---

```
[WARNING] *****************************************************************
[WARNING] * Your build is requesting parallel execution, but project      *
[WARNING] * contains the following plugin(s) that have goals not marked   *
[WARNING] * as @threadSafe to support parallel building.                  *
[WARNING] * While this /may/ work fine, please look for plugin updates    *
[WARNING] * and/or request plugins be made thread-safe.                   *
[WARNING] * If reporting an issue, report it against the plugin in        *
[WARNING] * question, not against maven-core                              *
[WARNING] *****************************************************************
[WARNING] The following goals are not marked @threadSafe in [snip]:
[WARNING] org.eclipse.xtext:xtext-maven-plugin:2.10.0:generate
[WARNING] *****************************************************************
```
<small>[(full-width log snippet)](https://gist.github.com/ciis0/130aaae5249499843561ac7a8258b9b7)</small>

**Comment by cdietrich** [comment]
emf and xtext is not thread safe. and it is not possible to make it generally thread safe.

**Comment by ciis0** [comment]
Well, that's very bad news. :/

**Comment by ciis0** [comment]
is it technically or economically not possible?

**Comment by cdietrich** [comment]
the effort to find a "general" solution is really high and there might be usecases where it is not possible at all. so the answer is yes to both points

**Comment by cdietrich** [comment]
did you already optimize the singlethreaded perf as much as you can?

**Comment by ciis0** [comment]
Well, I am not sure how to optimize the Maven build process further without parallelizing.

But do EMF and Xtext itself really need to be thread-safe, isn't it "enough" that the plugin handles EMF and Xtext in a thread-safe way?

Just to emphasize: I'm talking about the generation itself, not the generated code.
(I am unsure if I expressed my self well enough.)

**Comment by kthoms** [comment]
We understood that you are talking about the generation process itself. And parallelizing is surely one thing which seems an obvious way to improve performance. But without profiling your process you are just guessing that you might improve here. The first thing we usually do is profiling and then deduce how we can improve performance. We do this often, and very depending on the project's nature the tweaks are different.

Besides parallizing the Maven plugin's execution there is also the option to parallelize the generation process internally. But this requires also that you really know what you are doing. EMF is non thread-safe without any chance to do something about this. However, performance is important for us and our customers and there is always something what you can do about that. What exactly this means is project specific.

But to come to your original question: I don't think that the plugin can be changed to make EMF & Xtext thread-safe "enough". Otherwise, contributions welcome ;-)

**Comment by ciis0** [comment]
Okay, thank you so far. :)

**Comment by pfink** [comment]
Hi all,

I think the discussion got a slightly wrong spin here. Thread-safety doesn't mean that the plugin has to run faster than before or that it should be able to execute things in parallel. Thread-safety means that it does not produce wrong results (which it currently does) when executed in a multi-threaded build.

The xtext plugin is just one plugin out of many. That's why I think it's essential to make it thread-safe, so that it does not make multi-threaded build impossible at all as soon at it is used. There are many other plugins that can take advantage out of multi-threaded builds and the xtext-maven plugin should not prevent them from doing so.

I referenced a simple PR that ensures thread-safety for this plugin :)

**Comment by cdietrich** [comment]
i am fine with mark the plugin as not to support multithreading by adding the look
but i find it wired that the maven guys chose this model of opting out from multi threading instead of opting in.

**Comment by cdietrich** [comment]
Fixed in 2.15

**Comment by vijay44** [comment]
Just to confirm, "Fixed in 2.15" means is parallel maven build possible with this plugin?

**Comment by cdietrich** [comment]
no parall support in the plugin.

**Comment by cdietrich** [comment]
=> nobody has really made use of it

**Comment by kthoms** [comment]
But it means that the plugin can be used in a parallel Maven build. It itself won't do the work in multiple threads. But it is thread-safe by the synchronization now.

**Comment by vijay44** [comment]
Thanks for the clarification
